CoreFoundation/Base.subproj/ForFoundationOnly.h

```diff
@@ -121,44 +142,53 @@ typedef struct {
     CFStringRef _appName;
 } _CFApplicationPreferences;
 
+CF_EXPORT Boolean _CFPreferencesGetBooleanValueWithValue(CFPropertyListRef _Nullable value, Boolean * _Nullable keyExistsAndHasValidFormat);
+
 _CF_EXPORT_SCOPE_END
 
 
 
-// ---- CFString material ----------------------------------------
+#pragma mark - CFString
 
 #include <CoreFoundation/CFStringEncodingExt.h>
 
-#define NSSTRING_BOUNDSERROR \
-    [NSException raise:NSRangeException format:@"%@: Range or index out of bounds", __CFExceptionProem((id)self, _cmd)]
+#define NSSTRING_BOUNDSERROR do {\
+    _CFThrowFormattedException((CFStringRef)NSInvalidArgumentException, CFSTR("%@: Range or index out of bounds"), __CFExceptionProem(self, _cmd));\
+} while (0)
 
-#define NSSTRING_RANGEERROR(range, len) \
-    [NSException raise:NSRangeException format:@"%@: Range {%lu, %lu} out of bounds; string length %lu%s", __CFExceptionProem((id)self, _cmd), (unsigned long)range.location, (unsigned long)range.length, (unsigned long)len, ((range.length == __kCFStringInlineBufferLength) ? " (Note that the indicated range may be smaller than the original range passed to the API)" : "")]
+#define NSSTRING_RANGEERROR(range, len) do { \
+    _CFThrowFormattedException((CFStringRef)NSInvalidArgumentException, CFSTR("%@: Range {%lu, %lu} out of bounds; string length %lu%s"), __CFExceptionProem((id)self, _cmd), (unsigned long)range.location, (unsigned long)range.length, (unsigned long)len, ((range.length == __kCFStringInlineBufferLength) ? " (Note that the indicated range may be smaller than the original range passed to the API)" : ""));\
+} while (0)
 
-#define NSSTRING_INDEXERROR(index, len) \
-    [NSException raise:NSRangeException format:@"%@: Index %lu out of bounds; string length %lu", __CFExceptionProem((id)self, _cmd), (unsigned long)index, (unsigned long)len]
+#define NSSTRING_INDEXERROR(index, len) do { \
+    _CFThrowFormattedException((CFStringRef)NSInvalidArgumentException, CFSTR("%@: Index %lu out of bounds; string length %lu"), __CFExceptionProem((id)self, _cmd), (unsigned long)index, (unsigned long)len);\
+} while(0)
 
 // This can be made into an exception for post-10.10 apps
 #define NSSTRING_POSSIBLE_RANGEERROR(range, len)     \
-    if (__CFStringNoteErrors()) {       \
+    do {       \
         static bool warnonce = false;   \
         if (!warnonce) {                \
             warnonce = true;            \
             CFLog(kCFLogLevelWarning, CFSTR("*** %@: Range {%lu, %lu} out of bounds; string length %lu. This will become an exception for apps linked after 10.10 and iOS 8. Warning shown once per app execution."), __CFExceptionProem((id)self, _cmd), (unsigned long)range.location, (unsigned long)range.length, (unsigned long)len);        \
     }   \
-}
+} while(0)
```

Comment:
Nit: the indentation is off here.

Reply:
I'm gonna note these down and work on them later, for internal back-and-forth merging reasons.